src/mon/OSDMonitor.cc

```diff
@@ -906,12 +906,7 @@ void OSDMonitor::update_from_paxos(bool *need_bootstrap)
       if (state & CEPH_OSD_UP) {
 	// could be marked up *or* down, but we're too lazy to check which
 	last_osd_report.erase(osd);
-      }
-    }
-    for (auto [osd, weight] : inc.new_weight) {
-      if (weight == CEPH_OSD_OUT) {
-        // manually marked out, so drop it
-        osd_epochs.erase(osd);
+	osd_epochs.erase(osd);
```

Comment:
@Matan-B @athanatos Sorry for a late comment. We just had a customer that had stepped on this issue and noticed this PR, and now I am trying to understand the fix. And the comment above ("could be marked up *or* down, but we're too lazy to check which") makes we wonder won't we have cases when an osd is removed from `osd_epochs` when it is actually up?

Reply:
OSDMonitor marks OSD as down by updating its pending_inc.new_state to CEPH_OSD_UP, grep for `= CEPH_OSD_UP`. I agree this is confusing and probably worth revisiting out of scope of this PR.

Reply:
I see that e.g. in `OSDMap::Incremental::get_net_marked_down` when checking for `CEPH_OSD_UP` flag, it also checks the previous state to decide if it is marked down or up.

Anyway, I just made this experiment:
```
ceph osd down 0 
```
The mon set it down, the osd detected it is down in the osdmap and "reboot" and became up again in the osdmap, but now `ceph report` shows that the osd.0 is missing in `osd_epochs`. Triggering mon reelection makes the osd.0 back in `osd_epoch`.

Reply:
Marking an OSD down is not the same as actually shutting it down.

When marking the OSD down, it will get erased from osd_epochs. However, the OSD is still running and continues to send heartbeats. On the next heartbeat/beacon sent, the OSD will be inserted back to osd_epochs, regardless of mon's reelection.
You can test this by setting a short `osd_beacon_report_interval`.
If I didn't understand you correctly, please submit a tracker and we can continue the discussion there, Thanks!

Reply:
Ah, indeed the osd was reinserted eventually. I was not aware the default value for `osd_beacon_report_interval` was so large (5 minutes). I am still not sure though the behavior is entirely correct, though the issue seems very minor now. I will think if I want to open a ticket for this.

Reply:
Sounds good! 
btw, issues around `osd_epochs` can be fixed by simply restarting the leader mon since "stale" data will be removed on restart.

Reply:
> Ah, indeed the osd was eventually reinserted. I didn't know the default was so big (5 minutes). I'm still not sure if the behavior is entirely correct, although the issue seems minor right now. I'd consider whether or not to issue a ticket for this.`osd_beacon_report_interval`

hi @sawdust, How much could you change osd_beacon_report_interval to reproduce the problem?